sql/sql_class.h

```diff
@@ -3442,14 +3463,16 @@ class THD : public MDL_context_owner,
 
     @param new_db     the new database name.
 
-    @note This operation just sets {db, db_length}. Switching the current
-    database usually involves other actions, like switching other database
-    attributes including security context. In the future, this operation
-    will be made private and more convenient interface will be provided.
+    @note This operation just sets {db, db_length} and updated db_metadata
+    string for the thd. Switching the current database usually involves other
+    actions, like switching other database attributes including security
+    context. In the future, this operation will be made private and more
+    convenient interface will be provided.
   */
   void reset_db(const LEX_CSTRING &new_db) {
     m_db.str = new_db.str;
     m_db.length = new_db.length;
+    set_db_metadata();
```

Comment:
In running the test rpl_nogtid.rpl_stm_lcase_tblnames with debug builds, I am seeing a recursive lock error on LOCK_thd_data.

```
safe_mutex: Trying to lock mutex at /home/herman/rocks-mysql/8.0/sql/dd/dd_kill_immunizer.h, line 49, when the mutex was already locked at /home/herman/rocks-mysql/8.0/sql/sql_view.c
c, line 1301 in thread T@11

Thread 42 "mysqld" received signal SIGABRT, Aborted.
[Switching to Thread 0x7ffff5b40700 (LWP 843320)]
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7a21935 in __GI_abort () at abort.c:90
#2  0x00000000041f1473 in safe_mutex_lock (mp=0x7fffc342a0e0, try_lock=false, file=0x5868a23 "/home/herman/rocks-mysql/8.0/sql/dd/dd_kill_immunizer.h", line=49)
    at /home/herman/rocks-mysql/8.0/mysys/thr_mutex.cc:92
#3  0x000000000279680a in my_mutex_lock (mp=0x7fffc3407e30, file=0x5868a23 "/home/herman/rocks-mysql/8.0/sql/dd/dd_kill_immunizer.h", line=49)
    at /home/herman/rocks-mysql/8.0/include/thr_mutex.h:180
#4  0x0000000002794a21 in inline_mysql_mutex_lock (that=0x7fffc3407e30, src_file=0x5868a23 "/home/herman/rocks-mysql/8.0/sql/dd/dd_kill_immunizer.h", src_line=49)
    at /home/herman/rocks-mysql/8.0/include/mysql/psi/mysql_mutex.h:250
#5  0x0000000002796c71 in Mutex_lock::Mutex_lock (this=0x7ffff5b37430, mutex=0x7fffc3407e30, src_file=0x5868a23 "/home/herman/rocks-mysql/8.0/sql/dd/dd_kill_immunizer.h", 
    src_line=49) at /home/herman/rocks-mysql/8.0/include/mutex_lock.h:42
#6  0x0000000003ed0927 in dd::DD_kill_immunizer::DD_kill_immunizer (this=0x7ffff5b375e8, thd=0x7fffc3407000) at /home/herman/rocks-mysql/8.0/sql/dd/dd_kill_immunizer.h:49
#7  0x0000000003e97aaa in dd::Transaction_ro::Transaction_ro (this=0x7ffff5b375a0, thd=0x7fffc3407000, isolation=enum_tx_isolation::ISO_READ_COMMITTED)
    at /home/herman/rocks-mysql/8.0/sql/dd/impl/transaction_impl.h:134
#8  0x0000000003fd2d2d in dd::cache::Storage_adapter::get<dd::Item_name_key, dd::Schema> (thd=0x7fffc3407000, key=..., isolation=enum_tx_isolation::ISO_READ_COMMITTED, 
    bypass_core_registry=false, object=0x7ffff5b37670) at /home/herman/rocks-mysql/8.0/sql/dd/impl/cache/storage_adapter.cc:146
#9  0x0000000003fce600 in dd::cache::Shared_dictionary_cache::get_uncached<dd::Item_name_key, dd::Schema> (this=0x6ade8a0 <dd::cache::Shared_dictionary_cache::instance()::s_cache>, 
    thd=0x7fffc3407000, key=..., isolation=enum_tx_isolation::ISO_READ_COMMITTED, object=0x7ffff5b37670)
    at /home/herman/rocks-mysql/8.0/sql/dd/impl/cache/shared_dictionary_cache.cc:113
#10 0x0000000003fce56b in dd::cache::Shared_dictionary_cache::get<dd::Item_name_key, dd::Schema> (this=0x6ade8a0 <dd::cache::Shared_dictionary_cache::instance()::s_cache>, 
    thd=0x7fffc3407000, key=..., element=0x7ffff5b37778) at /home/herman/rocks-mysql/8.0/sql/dd/impl/cache/shared_dictionary_cache.cc:98
#11 0x0000000003ea0af2 in dd::cache::Dictionary_client::acquire<dd::Item_name_key, dd::Schema> (this=0x7fffc343b000, key=..., object=0x7ffff5b37810, local_committed=0x7ffff5b3780f, 
    local_uncommitted=0x7ffff5b3780e) at /home/herman/rocks-mysql/8.0/sql/dd/impl/cache/dictionary_client.cc:869
#12 0x0000000003e9aab2 in dd::cache::Dictionary_client::acquire<dd::Schema> (this=0x7fffc343b000, object_name="bug_37656", object=0x7ffff5b37968)
    at /home/herman/rocks-mysql/8.0/sql/dd/impl/cache/dictionary_client.cc:1214
#13 0x00000000028a2ced in THD::set_db_metadata (this=0x7fffc3407000) at /home/herman/rocks-mysql/8.0/sql/sql_class.cc:588
#14 0x00000000028df489 in THD::reset_db (this=0x7fffc3407000, new_db=...) at /home/herman/rocks-mysql/8.0/sql/sql_class.h:3475
#15 0x0000000002a9c1a6 in parse_view_definition (thd=0x7fffc3407000, view_ref=0x7fffc340bec8) at /home/herman/rocks-mysql/8.0/sql/sql_view.cc:1302
#16 0x0000000002878d81 in open_table (thd=0x7fffc3407000, table_list=0x7fffc340bec8, ot_ctx=0x7ffff5b38dd0) at /home/herman/rocks-mysql/8.0/sql/sql_base.cc:3253
#17 0x000000000287ebbf in open_and_process_table (thd=0x7fffc3407000, lex=0x7fffc3431000, tables=0x7fffc340bec8, counter=0x7fffc3431058, prelocking_strategy=0x7ffff5b38e98, 
    has_prelocking_list=false, ot_ctx=0x7ffff5b38dd0) at /home/herman/rocks-mysql/8.0/sql/sql_base.cc:5012
#18 0x000000000287da15 in open_tables (thd=0x7fffc3407000, start=0x7ffff5b38ea8, counter=0x7fffc3431058, flags=0, prelocking_strategy=0x7ffff5b38e98)
    at /home/herman/rocks-mysql/8.0/sql/sql_base.cc:5658
#19 0x00000000028820b5 in open_tables_for_query (thd=0x7fffc3407000, tables=0x7fffc340bec8, flags=0) at /home/herman/rocks-mysql/8.0/sql/sql_base.cc:6535
#20 0x00000000029ce9cc in Sql_cmd_dml::prepare (this=0x7fffc340c668, thd=0x7fffc3407000) at /home/herman/rocks-mysql/8.0/sql/sql_select.cc:384
#21 0x00000000029cf49b in Sql_cmd_dml::execute (this=0x7fffc340c668, thd=0x7fffc3407000) at /home/herman/rocks-mysql/8.0/sql/sql_select.cc:533
#22 0x000000000295fd86 in mysql_execute_command (thd=0x7fffc3407000, first_level=true, last_timer=0x7ffff5b3d6c0) at /home/herman/rocks-mysql/8.0/sql/sql_parse.cc:4668
#23 0x0000000002955e1b in mysql_parse (thd=0x7fffc3407000, parser_state=0x7ffff5b3ddc8, force_primary_storage_engine=false, last_timer=0x7ffff5b3d6c0)
    at /home/herman/rocks-mysql/8.0/sql/sql_parse.cc:5423
#24 0x0000000002952954 in dispatch_command (thd=0x7fffc3407000, com_data=0x7ffff5b3e2e8, command=COM_QUERY) at /home/herman/rocks-mysql/8.0/sql/sql_parse.cc:1817
#25 0x0000000002954ea7 in do_command (thd=0x7fffc3407000) at /home/herman/rocks-mysql/8.0/sql/sql_parse.cc:1273
#26 0x0000000002b3dc99 in handle_connection (arg=0x7fffea80f960) at /home/herman/rocks-mysql/8.0/sql/conn_handler/connection_handler_per_thread.cc:334
#27 0x00000000049db64e in pfs_spawn_thread (arg=0x7fffe5264220) at /home/herman/rocks-mysql/8.0/storage/perfschema/pfs.cc:2851
#28 0x00007ffff71896b6 in start_thread (arg=0x7ffff5b40700) at pthread_create.c:465
#29 0x00007ffff7b12ebf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

Reply:
Moving set_db_metadata() out of reset_db() and outside the lock seems to fix this issue. Would it make sense to load the db_metadata() information lazily when it is needed during gen_trx_metadata() and cache it until db is reset or changed?